app/controllers/auth/actions/AuthAction.scala

```diff
@@ -0,0 +1,210 @@
+          def newRequest: AuthenticatedRequest[A] =
+            requests.AuthenticatedRequest(
+              userId = externalId,
+              agentRef = agentRef,
+              saUtr = saUtr.map(SaUtr),
+              nino = nino.map(Nino(_)),
+              isAgentActive = isAgentActive,
+              confidenceLevel = confidenceLevel,
+              credentials = credentials,
+              request = request
+            )
+          (agentRef.isDefined, isAgentActive) match {
+            case (true, false) =>
+              Future.successful(Left(Redirect(controllers.routes.ErrorController.notAuthorised)))
+            case (true, true)  =>
+              agentTokenCheck(request, newRequest)
+            case _             =>
+              pertaxAuthService.authorise[A, Request[A]](request).map {
+                case Some(r) => Left(r)
+                case _       => Right(newRequest)
+              }
+          }
+        case _                                                                                              => throw new RuntimeException("Can't find credentials for user")
+      } recover {
+      case _: NoActiveSession        =>
+        lazy val ggSignIn    = appConfig.loginUrl
+        lazy val callbackUrl = appConfig.loginCallback
+        Left(
+          Redirect(
+            ggSignIn,
+            Map(
+              "continue_url" -> Seq(callbackUrl),
+              "origin"       -> Seq(appConfig.appName)
+            )
+          )
+        )
+      case _: InsufficientEnrolments => Left(Redirect(controllers.routes.ErrorController.notAuthorised))
+    }
+
+  private def agentInfo(enrolments: Set[Enrolment]): (Option[Uar], Boolean) =
+    enrolments
+      .find(_.key == "IR-SA-AGENT")
+      .map { enrolment =>
+        val d = enrolment.identifiers
+          .find(id => id.key == "IRAgentReference")
+          .map(key => Uar(key.value))
+        Tuple2(d, enrolment.isActivated)
+      }
+      .getOrElse(Tuple2(None, false))
+
+  private def citizenDetailsCheck[A](request: AuthenticatedRequest[A])(implicit
+    hc: HeaderCarrier
+  ): Future[AuthenticatedRequest[A]] =
+    (request.nino, request.saUtr, request.isAgent) match {
+      case (Some(nino), None, false) =>
+        getSAUTRFromCitizenDetails(nino).map {
+          case retrievedSAUtr @ Some(_) => request.copy(saUtr = retrievedSAUtr)
+          case None                     => request
+        }
+      case _                         => Future.successful(request)
+    }
+
+  private def getSAUTRFromCitizenDetails(nino: Nino)(implicit hc: HeaderCarrier): Future[Option[SaUtr]] =
+    citizenDetailsService.getMatchingDetails(nino.nino).map {
+      case SucccessMatchingDetailsResponse(matchingDetails) =>
+        matchingDetails.saUtr match {
+          case Some(_) => matchingDetails.saUtr
+          case _       => None
+        }
+      case _                                                => None
```

Comment:
You can't do that. You cannot recover from a server error by ignoring the utr.
You can have None only for a 404 response. If 404 is returned if no utr can be found

Reply:
This is what it does in main. However, you make a good point. I can change it so that it checks explicitly for 404 response and otherwise throws the exception back. Slack when you get in and we can discuss

Reply:
I've made changes. Not perfect as I don't really know why the connector needs to return an either which complicates the code a bit but it works now: any errors will redirect to error page.

Reply:
The old code needs a bit refactoring to better integrate with Either. You can do without those MatchingDetailsResponse, UpstreamErrorResponse will do just just fine. Use the service convert the 404 and parse the utr.
See next commit.

Reply:
That's much cleaner, thanks.